### DISCUSSION TITLE
What is the proper place to custom map/keybindings?

### DISCUSSION THREAD

**Discussion by SoaresAlisson** [R-nvim/R.nvim#219] [Q&A]
Thanks for this package!
I created my custom map/keybindings to some functions when developing my R packages. I'm using Lazyvim and placed it in `keymaps.lua` . It works, but I don't know if it is the proper way to do it, since it works even when no .R, .Rmd file is open. Is this the proper way to use custom keybindings, placing  it in `keymaps.lua` ?

```lua
local keymap = vim.api.nvim_set_keymap
keymap("n", "<A-r>dl", ":RSend devtools::load_all()<cr>", { desc = "R devtools load all" })
keymap("n", "<A-r>dd", ":RSend devtools::document()<cr>", { desc = "R devtools document" })
keymap("n", "<A-r>db", ":RSend devtools::build(binary=FALSE)<cr>", { desc = "R devtools build package" })
keymap("n", "<A-r>di", ":RSend devtools::install()<cr>", { desc = "R devtools install package" })
```

Thanks!

**Comment by PMassicotte** [comment]
I use `ftplugin` for my custom keymaps.

**Comment by jalvesaq** [comment]
This is a good question, and the use `ftplugn` (and, perhaps, `after/ftplugin`) directory is a good solution. Our mappings use the `<Plug>` labels which were optimized by @BlueDrink9 (https://github.com/jalvesaq/Nvim-R/commit/fe5edb9b8737766f24cdba3b8a6b086480572fa3). So, perhaps you can create global maps for the `<Plug>` labels...

**Comment by BlueDrink9** [comment]
In this case I don't think the plug mappings apply, because you're specifically
sending an argument to the command.

The proper way to do what you want in Neovim is to use buffer local mappings, meaning they only apply in R buffers and they aren't mapped in others.

Since you're using LazyVim, follow the [LazyVim example](https://www.lazyvim.org/extras/lang/r#rnvim)

**Comment by jalvesaq** [comment]
Thanks, @BlueDrink9! As in LazyVim, I moved the examples in the README of how to set keymaps to the `on_filetype` hook.

**Comment by TymekDev** [comment]
Personally, I have just opted for using `keys` in [lazy.nvim](https://github.com/folke/lazy.nvim) plugin spec:
```lua
local commands = {
  "devtools::load_all()",
}

---@module "lazy"
---@type LazySpec
return {
  "R-nvim/R.nvim",
  ft = "r",
  keys = {
    {
      "<Leader>re",
      function()
        vim.ui.select(commands, { prompt = "Execute in R" }, function(cmd)
          if cmd ~= nil then
            vim.cmd.RSend(cmd)
          end
        end)
      end,
      ft = "r",
      desc = "Execute one of a predefined commands in R (via R.nvim)",
    },
  },
  opts = {},
}
```
